File: 2007/CCC '07 S5 - Bowling for Numbers.cpp
#include <iostream>
#include <vector>
using namespace std;

int solve(int i, int ball, vector<vector<int>>& cache, vector<int>& scores, int W) {
    if (cache[i][ball] != -1) {
        return cache[i][ball];
    }
    if (ball == 0 || i >= scores.size()) {
        return 0;
    }
    cache[i][ball] = max(scores[i] + solve(i + W, ball - 1, cache, scores, W), solve(i + 1, ball, cache, scores, W));
    return cache[i][ball];
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int T;
    cin >> T;

    while (T--) {
        int N, K, W;
        cin >> N >> K >> W;

        vector<int> arr(N + W);
        for (int i = 0; i < N; i++) {
            cin >> arr[i];
        }
        for (int i = N; i < N + W; i++) {
            arr[i] = 0;
        }

        N += W;

        vector<int> scores(N - W + 1);
        scores[0] = 0;
        for (int i = 0; i < W; i++) {
            scores[0] += arr[i];
        }
        for (int i = 1; i < scores.size(); i++) {
            scores[i] = scores[i - 1] + arr[i + W - 1] - arr[i - 1];
        }

        vector<vector<int>> cache(N + 1, vector<int>(K + 1, -1));

        cout << solve(0, K, cache, scores, W) << endl;
    }

    return 0;
}

File: 2007/CCC '07 S5 - Bowling for Numbers.py
# DP, for each bowling pin, we can either hit it along with the next W-1 pins, or leave it
# precompute the score gained by throwing at bowling ball at each pin in O(n)

# top-down approach, too slow/memory consuming for last test case
# check c++ code for working solution
import sys

sys.setrecursionlimit(30001)
input = sys.stdin.readline


def solve(i, ball):
    if cache[i][ball] != -1:
        return cache[i][ball]
    if ball == 0 or i >= len(scores):  # out of bowling balls or out of bounds
        return 0

    cache[i][ball] = max(scores[i] + solve(i + W, ball - 1), solve(i + 1, ball))
    return cache[i][ball]


for _ in range(int(input())):  # for each test case
    N, K, W = map(int, input().split())
    arr = [int(input()) for _ in range(N)] + [0] * W  # pad arrays with 0s
    N += W

    # precompute score gained for hitting the pins from score[i] to score[i+W-1]
    scores = [0] * (N - W + 1)
    scores[0] = sum(arr[:W])
    for i in range(1, len(scores)):
        scores[i] = scores[i - 1] + arr[i + W - 1] - arr[i - 1]

    cache = [[-1] * (K + 1) for _ in range(N + 1)]
    print(solve(0, K))
